services/cd-service/pkg/service/git.go

```diff
@@ -153,10 +157,95 @@ func (s *GitServer) GetCommitInfo(ctx context.Context, in *api.GetCommitInfoRequ
 	}, nil
-}
\ No newline at end of file
+}
+
+func (s *GitServer) GetEvents(ctx context.Context, fs billy.Filesystem, hash string, commitPath string) ([]*api.Event, error) {
+	errorTemplate := func(message string, err error) error {
+		return fmt.Errorf("could not get events for hash '%s' in path '%s' - error %w", hash, commitPath, err)
+	}
+	var result []*api.Event = nil
+	allEventsPath := fs.Join(commitPath, "events")
+	potentialEventDirs, err := fs.ReadDir(allEventsPath)
+	if err != nil {
+		return nil, errorTemplate("could not read events directory ", err)
+	}
+	for i := range potentialEventDirs {
+		oneEventDir := potentialEventDirs[i]
+		if oneEventDir.IsDir() {
+			fileName := oneEventDir.Name()
+			rawUUID, err := timeuuid.ParseUUID(fileName)
+			if err != nil {
+				return nil, errorTemplate(fmt.Sprintf("could not read event, because it is not a UUID '%s'", fileName), err)
+			}
+
+			var event *api.Event = nil
+			event, err = s.ReadEvent(ctx, fs, fs.Join(allEventsPath, oneEventDir.Name()), rawUUID)
+			if err != nil {
+				return nil, errorTemplate("could not read event", err)
+			}
+			result = append(result, event)
+		}
+	}
+	sort.Slice(result, func(i, j int) bool {
+		return result[i].CreatedAt.AsTime().UnixNano() < result[j].CreatedAt.AsTime().UnixNano()
+	})
+	return result, nil
+}
+
+func (s *GitServer) ReadEvent(ctx context.Context, fs billy.Filesystem, eventPath string, eventId timeuuid.UUID) (*api.Event, error) {
+	errorTemplate := func(message string, err error) error {
+		return fmt.Errorf("could not read events - error %w", err)
+	}
+	eventTypePath := fs.Join(eventPath, "eventType")
+	eventTypeRaw, err := util.ReadFile(fs, eventTypePath) // full path: commits/<commitHash2>/<commitHash38>/events/<eventUUID>/eventType
+	if err != nil {
+		return nil, errorTemplate(fmt.Sprintf("could not read eventType %s", eventTypePath), err)
+	}
+	var eventType = string(eventTypeRaw)
+
+	if eventType == eventmod.NewReleaseEventName {
+		eventEnvsPath := fs.Join(eventPath, "environments")
+
+		potentialEnvironmentDirs, err := fs.ReadDir(eventEnvsPath)
+		if err != nil {
+			return nil, errorTemplate(fmt.Sprintf("could not read events environments %s", eventEnvsPath), err)
+		}
+
+		var envs []string = nil
+		for i := range potentialEnvironmentDirs {
+			envDir := potentialEnvironmentDirs[i]
+			fileName := envDir.Name()
+			if envDir.IsDir() {
+				envs = append(envs, fileName)
+			} else {
+				logger.FromContext(ctx).Sugar().Warnf(
+					"found entry in %s that was not an environment directory", fs.Join(eventEnvsPath, fileName))
+			}
+		}
+
+		var result = &api.Event{
+			CreatedAt: uuid.GetTime(&eventId),
+			EventType: &api.Event_CreateReleaseEvent{
+				CreateReleaseEvent: &api.CreateReleaseEvent{
+					EnvironmentNames: envs,
+				},
+			},
+		}
+		return result, nil
+	}
+	err = fmt.Errorf("could not read event, did not recognize event type '%s'", eventType)
+	logger.FromContext(ctx).Sugar().Warnf("%v", err)
+	return nil, err
```

Comment:
Do not log & return errors, this is an antipattern

Reply:
fair point, fixed